Several improvements to SV contig alignment configuration picker

SHuang-Broad opened the pull request:
Fix a particular "bug" observed that caused us dozens of variants:

When an assembly contig has a unique region covered by one or more MQ 0 alignments, the current configuration picker classifies such contig as "ambiguous". 
This PR fix this problem by implementing a tie breaker saying that for such contigs, we prefer the unique configuration with no MQ 0 alignments; that is, this tie breaker only saves contigs have one unique configuration with all non-MQ-0 mappings.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java:
```diff
@@ -78,4 +91,26 @@ private static void evaluateIntervalsAgainstTruth(final List<SVInterval> assembl
+                throw new UserException("Can't read nonCanonicalContigNamesFile file "+nonCanonicalContigNamesFile, ioe);
+            }
+        } else {
+            final List<String> first22ChromosomesNum = IntStream.range(0, 23).mapToObj(String::valueOf).collect(Collectors.toList());
```
Can we not do this? It's better to require that either you have the file or you don't. If you don't have the file just treat everything as canonical. I don't want to hard code any chromosome names in the system, and I definitely don't want the tool to have any opinion on whether hg19/b37 chromosome names have "chr" in them. I know the code has been there for a while but somehow I don't remember reviewing this; maybe I just missed it in the original large pull request.

SHuang-Broad replied:
removed

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
"that have"

SHuang-Broad replied:
fixed

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
Do we need to split the rdd, process each half separately, and then union? That seems really inefficient. Can't you write this as a series of maps and filters?

SHuang-Broad replied:
updated

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
This line of code appears in `hasOneAndOnlyOneAllNonZeroMQs` as well, maybe extract a method.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
Add message to exception.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
Why are these called `savedOnes` and `unsavedOnes`? The names don't seem accurate to me (we end up keeping all of the `unsavedOnes`, no?

cwhelan commented:
Is it guaranteed that one of the configurations won't include any of the MQ0 regions? Why is that?

cwhelan commented:
When you say this "caused us dozens of variants" do you mean true positives or false negatives?

SHuang-Broad commented:
@cwhelan Thanks for the review! And I apologize for not clearly stating what problem is getting fixed here.

I've addressed the comments in separate commit, changed the implementation, made more improvements that were discovered while reviewing the variants 
(https://github.com/SHuang-Broad/GATK-SV-callset-regressionTest/tree/master/Evaluation/Analysis/masterVSfeature/notes.xlsx)
The implemented fixes are:
* for removing the hard-coded/explicit mentioning of "chr" in non-canonical versions, it is now fixed in 5eff782e4d582d516004fba2cee7535d984b1540
* for contigs whose alignments paint ambiguous picture, i.e. multiple alignment configurations offer equally good explanation:
	1. if only one configuration has all alignment with MQ above a specified threshold, it is favored; this is implemented in ecc31f5fbec4e524b401fc9474a3a1b7ab08c561
	2. if one configuration has alignment to non-canonical chromosome that explains the contig better than would-be-event-inducing mappings to canonical chromosomes, the canonical mappings are saved but the better non-canonical mappings are saved as SA tag as in SAM spec, and the VCF record produced is annotated accordingly; this is implemented in 65cdb523a2f9fa2026334713fed45381d76ffc82
* fixed a bug where sometimes an assembly contig as several alignments, only one of which has non-mediocre MQ but at the sametime this alignment contains a large gap, such contigs were previously incorrectly filtered away, they are now salvaged by commit b6b2f197b112981e00efd9d415f010c024d31b36

So, for the FN variants (FN in the sense that they are captured in the stable version of our interpretation tool but now goes missing in the experimental interpretation tool)
that were curated in the above-mentioned review, only the following ones are not salvaged, with plans or comments attached

```
asm012854:tig00000	missing	classified as "incomplete"; fixable by finishing the last TODO in AssemblyContigAlignmentSignatureClassifier (same problem as face by group represented by asm002398:tig00001)
asm014580:tig00018	missing	classified as "incomplete"; fixable by finishing the last TODO in AssemblyContigAlignmentSignatureClassifier (same problem as face by group represented by asm002398:tig00001)
asm008185:tig00000	missing	classified as "ambiguous"; unsure of how to deal with a general rule--exact same part of read coverred by one mapping to canonical, one to non-canonical (better AS); if save the canonical, a deletion would be called
asm018220:tig00004	missing	classified as "ambiguous"; salvagable if re-analysed through old interpretation tool
asm026229:tig00000	missing	classified as "ambiguous"; fixable if implement a special rule making all split alignments on non-canonical chromosomes as "bad"
```

They are relatively simple fixes so I clumped them into a single PR.
If you prefer to review them as separate PR's. I can do that too.
Thanks!

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigWithFineTunedAlignments.java:
```diff
@@ -29,20 +30,29 @@
+    public AssemblyContigWithFineTunedAlignments(@Nonnull final AlignedContig contig,
+                                                 @Nonnull final List<String> insertionMappings,
+                                                 final AlignmentInterval goodMappingToNonCanonicalChromosome) {
+        this(Utils.nonNull(contig), Utils.nonNull(insertionMappings),
```
Do you need both calls to `Utils.nonNull` and the annotations in the method signature?

SHuang-Broad replied:
maybe I'm just paranoid.... removed the annotation

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -461,7 +509,42 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
+     *          otherwise the non-canonical chromosome mapping
+     */
+    @VisibleForTesting
+    static AlignmentInterval specialChanelForSingleNonCanonicalMappings(final Set<String> canonicalChromosomes,
```
changed as suggested

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -170,28 +185,41 @@ public boolean equals(final Object o) {
+        if (alignedContig.alignmentIntervals.size() == 1) {
+            return Collections.singletonList(
+                    new GoodAndBadMappings(Collections.singletonList(alignedContig.alignmentIntervals.get(0)),
+                                           Collections.emptyList(), null)
```
ah, right! Thanks for catch that!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -461,7 +509,42 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
+     *  one configuration has split alignments to canonical chromosomes, hence indicating the SV, whereas
+     *  the other configuration has a single, often very good (or even better) alignment to a non-canonical chromosome.
+     * We send down the chimeric alignment configuration for inference, but notes down that a non-canonical chromosome
+     * in the reference bundle input could have already captured the SV on this sample.
```
corrected. Thanks for the explanation!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -461,7 +509,42 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
-    // one configuration with single good mapping to non-canonical chromosome,
-    // another configuration with split alignments to canonical chromosomes
+    /**
+     * There are locations on the non-canonical chromosomes of the HG38 reference that is similar to a location
```
corrected

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -24,8 +24,22 @@
+     * either:
+     *  has only 1 mapping, with MQ strictly above this threshold
+     * or:
+     *  has more than 1 mappings, but only 1 mapping has MQ strictly above this threshold and it has a large gap in it.
```
corrected

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -322,4 +368,30 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
+     *  hence remove the ambiguity.
+     */
+    @VisibleForTesting
+    static List<GoodAndBadMappings> breakTieByPreferringLessAlignments(
```
Sorry for not squashing the later bug fix commit here.
You are right, it is a bad choice of name for the method. 
But I did change the name as suggested. Thanks!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java:
```diff
@@ -116,20 +116,58 @@ private static boolean notDiscardForBadMQ(final AlignedContig contig) {
+     * After configuration scoring and picking, the original alignments can be classified as
+     * good and bad mappings:
+     * good: the ones present the picked configuration
+     * bad: the brings more noise than information,
```
corrected

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java:
```diff
@@ -85,4 +97,20 @@ private static void evaluateIntervalsAgainstTruth(final List<SVInterval> assembl
+            return allContigsStream.collect(Collectors.toCollection(LinkedHashSet::new));
+
+        try (final Stream<String> nonCanonical = Files.lines(IOUtils.getPath(( Utils.nonNull(nonCanonicalContigNamesFile) )))) {
+            final HashSet<String> toIgnore = nonCanonical.collect(Collectors.toCollection(HashSet::new));
```
removed. Thanks!
